### PR TITLE
Vuex

### PR DESCRIPTION
**Pull request by ezzatmohamed**

### CODE REVIEW

File: server/database/migrations/2020_12_15_120531_add_like_column_to_favorites.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('favorites', function (Blueprint $table) {
+            //
+        });
```
Comment (A-Elshorafa) [inline server/database/migrations/2020_12_15_120531_add_like_column_to_favorites.php]: drop the column if exists

File: server/nova-components/FavoritesGrid/resources/js/components/Tool.vue [hunk 2]
```diff
@@ -1,22 +1,25 @@
...
-
     </div>
 </template>
 
 
 <script>
-import {parseNovaApi} from './../../../../StoryBook/resources/js/helpers'
-
+import { mapState, mapActions} from 'vuex'
+import {selectInput} from './../../../../StoryBook/resources/js/components/inputs'
```
Comment (A-Elshorafa) [inline server/nova-components/FavoritesGrid/resources/js/components/Tool.vue]: Shouldn't to import component like this

File: server/nova-components/FavoritesGrid/resources/js/components/Tool.vue [hunk 3]
```diff
@@ -27,38 +30,36 @@ export default {
...
+                for( let key in this.columns)
+                    columnAttributes.push(key)    
+                return columnAttributes
+            },
+            ...mapState({
+                    favoritesList: state => state.favoriteModule.favoritesList,
+                    favoriteFilter: state => state.favoriteModule.favoriteFilter,
+                }),
+    },
+    components:{selectInput},
```
Comment (A-Elshorafa) [inline server/nova-components/FavoritesGrid/resources/js/components/Tool.vue]: Remove this please

File: server/nova-components/StoryBook/resources/js/vuex/modules/userModule/actions.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+const actions = {
+    
+    retrieveUsers({commit}){
+        Nova.request()
+        .get('/nova-api/users')
+        .then(res=>{
+            const users = parseNovaApi(res,["id","name"])
+            commit('setUsersList',users)
+        })
```
Comment (A-Elshorafa) [inline server/nova-components/StoryBook/resources/js/vuex/modules/userModule/actions.js]: validate `res`

File: server/database/migrations/2020_12_15_120531_add_like_column_to_favorites.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        Schema::table('favorites', function (Blueprint $table) {
+            if (Schema::hasColumn('favorites', 'liked'))
+            {
+                Schema::table('favorites', function (Blueprint $table)
+                {
+                    $table->dropColumn('liked');
+                });
+            }
+        });
```
Comment (A-Elshorafa) [inline server/database/migrations/2020_12_15_120531_add_like_column_to_favorites.php]: Focus you duplicate this line `Schema::table('favorites', function (Blueprint $table)`

File: server/nova-components/FavoritesGrid/resources/js/components/Tool.vue [hunk 6]
```diff
@@ -27,38 +29,35 @@ export default {
...
+            ]
+        }
+    },
+    computed: {
+            columnAttribute(){
+                let columnAttributes = []
+                for( let key in this.columns)
+                    columnAttributes.push(key)    
+                return columnAttributes
+            },
```
Comment (A-Elshorafa) [inline server/nova-components/FavoritesGrid/resources/js/components/Tool.vue]: Check `Object.keys()`

File: server/nova-components/StoryBook/resources/js/vuex/modules/createFavoriteModule/actions.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    changeInfo({commit},payload){
+        if(payload)
+            commit('changeInfo',payload)
+        
+    },
+    changeError({commit},payload){
+        if(payload)
+            commit('changeError',payload)
+        
+    },
```
Comment (A-Elshorafa) [inline server/nova-components/StoryBook/resources/js/vuex/modules/createFavoriteModule/actions.js]: Why don't use them as Mutations directly with `mapMutations`

File: server/nova-components/StoryBook/resources/js/vuex/modules/favoriteModule/actions.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                {
+                    const favorites = parseNovaApi(res,payload.columnAttribute)
+                    
+                    if(Array.isArray(favorites))
+                    {
+                        favorites.forEach((data,i)=>{
+                          
+                          data['deleted_at'] = res && res.data && 
+                                              res.data.resources[i] && 
+                                              res.data.resources[i].softDeleted ? "Yes" : "No"
```
Comment (A-Elshorafa) [inline server/nova-components/StoryBook/resources/js/vuex/modules/favoriteModule/actions.js]: Should be at parseNovaApi helper

File: server/nova-components/StoryBook/resources/js/vuex/modules/favoriteModule/actions.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    retrieveFavorites({commit},payload){
+
+        let endpoint = '/nova-api/favorites?trashed=with'
+        if(payload && payload.filter)
+        {   
+            commit('setFavoriteFilter',payload.filter)
+            if(payload.filter == 1 )
+                endpoint += '&liked=1'
+            else if(payload.filter == 2)
+                endpoint += '&liked=0'
```
Comment (A-Elshorafa) [inline server/nova-components/StoryBook/resources/js/vuex/modules/favoriteModule/actions.js]: Check how to use nova filters

File: server/app/Nova/Favorite.php [hunk 10]
```diff
@@ -90,7 +91,12 @@ public function fields(Request $request)
 
     // Return all favorites for admin but for regulars,return only their favorites
     public static function indexQuery(NovaRequest $request, $query)
-    {
+    {  
+        if($request->has('liked'))
+        {
+            $query = $request->query('liked') == 1 ? $query->where('liked',true)
+                                                        : $query = $query->where('liked',false);
+        }
```
Comment (A-Elshorafa) [inline server/app/Nova/Favorite.php]: Check how to use nova filters

File: server/app/Nova/Filters/LikeFilter.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+     * Apply the filter to the given query.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Illuminate\Database\Eloquent\Builder  $query
+     * @param  mixed  $value
+     * @return \Illuminate\Database\Eloquent\Builder
+     */
+    public function apply(Request $request, $query, $value)
+    {
+        // $id = State::where('name','banned')->pluck('id')->first();
```
Comment (abdelrahmanna) [inline server/app/Nova/Filters/LikeFilter.php]: remove commented code

File: server/app/Nova/State.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+     * @var string
+     */
+    public static $model = \App\Models\State::class;
+
+    /**
+     * The single value that should be used to represent the resource when being displayed.
+     *
+     * @var string
+     */
+    public static $title = 'id';
```
Comment (abdelrahmanna) [inline server/app/Nova/State.php]: set title to name

File: server/app/Nova/Filters/FavoriteState.php [hunk 13]
```diff
@@ -29,13 +29,17 @@ public function apply(Request $request, $query, $value)
...
-                        ->orWhere(function($query) use ($ids,$no_state){
-                                if($no_state)
-                                    $query->doesntHave('states');
-                        });
+        return $query->where(function($query) use($ids,$no_state){
+                
+                // Get favorites with the specified states
+                $query->whereHas('states', function($query) use ($ids) {
+                            $query->whereIn('state_id',$ids);
+                     })->orWhere(function($query) use ($ids,$no_state){
```
Comment (abdelrahmanna) [inline server/app/Nova/Filters/FavoriteState.php]: remove $ids

**Pull request merged**